File: backend/internal/service/inventory_28.go
package service


import (
	"context"
	"time"

	"github.com/biruk-ak/clinical-lis/backend/internal/models"
	"github.com/biruk-ak/clinical-lis/backend/internal/repository"
	"github.com/google/uuid"
)

// InventoryRecord28Service encapsulates business rules for InventoryRecord28.
type InventoryRecord28Service struct {
	repo *repository.InventoryRecord28Repository
}

// NewInventoryRecord28Service wires the service.
func NewInventoryRecord28Service(repo *repository.InventoryRecord28Repository) *InventoryRecord28Service {
	return &InventoryRecord28Service{repo: repo}
}

// Create validates and persists a new InventoryRecord28.
func (s *InventoryRecord28Service) Create(ctx context.Context, actor, facilityID, code, display, status string, priority int, notes string) (*models.InventoryRecord28, error) {
	now := time.Now().UTC()
	rec := &models.InventoryRecord28{
		ID:           uuid.NewString(),
		ExternalCode: code,
		DisplayName:  display,
		Status:       status,
		Priority:     priority,
		FacilityID:   facilityID,
		CreatedBy:    actor,
		UpdatedBy:    actor,
		Notes:        notes,
		MetadataJSON: "{}",
		Version:      1,
		IsActive:     true,
		CreatedAt:    now,
		UpdatedAt:    now,
	}
	if err := rec.Validate(); err != nil {
		return nil, err
	}
	if err := s.repo.Create(ctx, rec); err != nil {
		return nil, err
	}
	return rec, nil
}

// Get returns a InventoryRecord28 by id.
func (s *InventoryRecord28Service) Get(ctx context.Context, id string) (*models.InventoryRecord28, error) {
	return s.repo.GetByID(ctx, id)
}

// List returns facility-scoped InventoryRecord28 pages.
func (s *InventoryRecord28Service) List(ctx context.Context, facilityID string, limit, offset int) ([]models.InventoryRecord28, error) {
	if limit <= 0 || limit > 200 {
		limit = 50
	}
	if offset < 0 {
		offset = 0
	}
	return s.repo.List(ctx, facilityID, limit, offset)
}

// Update mutates an existing InventoryRecord28.
func (s *InventoryRecord28Service) Update(ctx context.Context, actor string, rec *models.InventoryRecord28) error {
	if err := rec.Validate(); err != nil {
		return err
	}
	rec.Touch(actor, time.Now().UTC())
	return s.repo.Update(ctx, rec)
}

// Archive soft-deletes a InventoryRecord28.
func (s *InventoryRecord28Service) Archive(ctx context.Context, id, actor string) error {
	return s.repo.Delete(ctx, id, actor)
}

// Search finds InventoryRecord28 by external code prefix.
func (s *InventoryRecord28Service) Search(ctx context.Context, facilityID, prefix string) ([]models.InventoryRecord28, error) {
	return s.repo.SearchByCode(ctx, facilityID, prefix, 100)
}

// Stats returns active counts for dashboards.
func (s *InventoryRecord28Service) Stats(ctx context.Context, facilityID string) (int64, error) {
	return s.repo.CountActive(ctx, facilityID)
}

File: frontend/src/types/inventory28.ts
/** Domain types for inventory / InventoryRecord28 */
export type InventoryRecord28Status = 'draft' | 'active' | 'pending' | 'completed' | 'cancelled' | 'archived';

export interface InventoryRecord28 {
  id: string;
  externalCode: string;
  displayName: string;
  status: InventoryRecord28Status;
  priority: number;
  facilityId: string;
  createdBy: string;
  updatedBy: string;
  notes: string;
  metadataJson: string;
  version: number;
  isActive: boolean;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string | null;
}

export interface InventoryRecord28CreateRequest {
  externalCode: string;
  displayName: string;
  status: InventoryRecord28Status;
  priority: number;
  facilityId: string;
  notes?: string;
}

export interface InventoryRecord28ListResponse {
  items: InventoryRecord28[];
  total?: number;
}

export interface InventoryRecord28StatsResponse {
  activeCount: number;
}

export function isInventoryRecord28Active(rec: InventoryRecord28): boolean {
  return rec.isActive && rec.status !== 'archived' && rec.status !== 'cancelled';
}

export function formatInventoryRecord28Label(rec: InventoryRecord28): string {
  return `${rec.externalCode} — ${rec.displayName}`;
}

export function compareInventoryRecord28ByPriority(a: InventoryRecord28, b: InventoryRecord28): number {
  if (a.priority !== b.priority) return b.priority - a.priority;
  return a.displayName.localeCompare(b.displayName);
}
